### PR TITLE
New-CmAzDeployment first implementation

### LINKED ISSUE #91: General - Self-loading configuration files
Currently we have to remember which configuration file is associated with each CmdLet. However, we can simplify our deploy scripts considerably with a single deployment command and parameter files which indicate the appropriate provisioning mechanism. Example file named `mycomponent.yml`:

``` yaml
component: core.keys
# rest of config goes here
```

This could then be deployed using a much simpler command, along the lines of:

``` powershell
New-CmAzDeployment -SettingsFile mycomponent.yml
```

Within the command we include what is effectively a large switch statement:

``` powershell
Switch (settings.component)
{
  "core.keys" { New-CmAzCoreKeyVault -SettingsObject settings }
}
```

As well as simpler usage in deployment scripts, this change would lay the groundwork for being able to auto-deploy based on static settings files alone, massively reducing the need for scripting.

- [x] Add "component" to each settings file created via New-CmAzProject, with the appropriate value: Keys, VM etc.
- [x] Create new cmdlet called New-CmAzDeplotment, this should take the settings file as the parameter.
- [x] Create a switch statement that checks the value of the component property in the settings file, and calls the appropriate 
cmdlet. If the component cannot be found, throw an error highlighting this.
- [x] Write a pester test that determines that the correct cmdlet is called for each component. Test the output in the pipeline as well (Use WhatIf and WhatIfPreference).
- [x] Write a pester test checking that an error is thrown when the component is not found, test that the correct error message is displayed as well.

### PR DESCRIPTION
**Pull request by deepend-dev**
resolves #91

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -1,181 +1,177 @@
 parameters:
-- name: ReleaseNotes
```
Comment (tm-luke) [inline azure-pipelines.yml:2]: revert indentation

File: Tests/New-CmAzDeployment.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,39 @@
+Sync-CloudmarqueAzure
+
+$excludes = @("generators.yml", "naming.yaml")
+$resourceFiles = Get-ChildItem -Path "$PSScriptRoot\..\Cloudmarque.Azure\Resources\Project" -Recurse -Force -Exclude $excludes -File | Select-String "component" -List | Select-Object Filename, Path
+
+Describe "New-CmAzDeployment Tests" {
+
+    Context "Component identified successfully" {
```
Comment (tm-luke) [inline Tests/New-CmAzDeployment.Tests.ps1:8]: Cmdlet instead of component

File: Tests/New-CmAzDeployment.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+        It "should be able to call associated command for <_.Filename>" -TestCases $resourceFiles {
```
Comment (tm-luke) [inline Tests/New-CmAzDeployment.Tests.ps1:10]: the correctly associated command for <_.Filename>"

File: Tests/New-CmAzDeployment.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+            } | Should -Not -Throw "Please provide appropriate component value in settings file or object.."
+        }
+    }
+
+    Context "Component not identified" {
+
+        It "should throw error message" {
+            {
+                New-CmAzDeployment -SettingsFile "$PSScriptRoot\..\Cloudmarque.Azure\Resources\Project\security.yml" -whatif
```
Comment (tm-luke) [inline Tests/New-CmAzDeployment.Tests.ps1:35]: -WhatIf

File: Tests/New-CmAzDeployment.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        It "should be able to call associated command for <_.Filename>" -TestCases $resourceFiles {
+            {
+                $deployments = ""
+
+                if ($_.Filename -eq "virtualmachines.yml") {
+
+                    $username = ConvertTo-SecureString -String "testuser" -AsPlainText -Force
+                    $password = ConvertTo-SecureString -String (Get-Password -MinPasswordLength 20) -AsPlainText -Force
+
+                    $deployments = New-CmAzDeployment -SettingsFile $_.Path -LocalAdminUsername $username -LocalAdminPassword $password -whatif -verbose 4>&1 -Erroraction continue
```
Comment (tm-luke) [inline Tests/New-CmAzDeployment.Tests.ps1:19]: -WhatIf, -Verbose

File: Tests/New-CmAzDeployment.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                }
+                else {
+                    $deployments = New-CmAzDeployment -SettingsFile $_.Path -whatif -verbose 4>&1 -Erroraction continue
+                }
```
Comment (tm-luke) [inline Tests/New-CmAzDeployment.Tests.ps1:23]: Why are we using ErrorAction continue here?

File: Tests/New-CmAzDeployment.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+                $deployments.toString() | Should -Not -BeNull
+
+            } | Should -Not -Throw "Please provide appropriate component value in settings file or object.."
```
Comment (tm-luke) [inline Tests/New-CmAzDeployment.Tests.ps1:27]: Store message in a variable and reuse across tests

File: Cloudmarque.Azure/Public/Deploy/Security/security-centre/Set-CmAzSecurityCentre.ps1 [hunk 8]
```diff
@@ -40,7 +40,9 @@ function Set-CmAzSecurityCentre {
 
     try {
 
-        if ($PSCmdlet.ShouldProcess((Get-CmAzSubscriptionName), "Deploy Security Centre")) {
+        Get-InvocationInfo -CommandName $MyInvocation.MyCommand.Name
+
+		if ($PSCmdlet.ShouldProcess((Get-CmAzSubscriptionName), "Deploy Security Centre")) {
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/Security/security-centre/Set-CmAzSecurityCentre.ps1:45]: Is this correctly indented?

File: Cloudmarque.Azure/Public/Deploy/Security/managed-services/New-CmAzSecurityPartner.ps1 [hunk 9]
```diff
@@ -31,7 +31,9 @@
 	)
 	try {
 
-        if ($PSCmdlet.ShouldProcess((Get-CmAzSubscriptionName), "Deploy managed service security groups.")) {
+        Get-InvocationInfo -CommandName $MyInvocation.MyCommand.Name
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/Security/managed-services/New-CmAzSecurityPartner.ps1:34]: Is this correctly indented?

File: Cloudmarque.Azure/Public/Deploy/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1 [hunk 10]
```diff
@@ -39,7 +39,9 @@
 
     try {
 
-        if ($PSCmdlet.ShouldProcess((Get-CmAzSubscriptionName), "Deploy shared image gallery")) {
+        Get-InvocationInfo -CommandName $MyInvocation.MyCommand.Name
+
+		if ($PSCmdlet.ShouldProcess((Get-CmAzSubscriptionName), "Deploy shared image gallery")) {
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1:44]: Is this correctly indented?

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/Set-CmAzIaasUpdateManagement.ps1 [hunk 11]
```diff
@@ -33,6 +33,8 @@ function Set-CmAzIaasUpdateManagement {
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/IaaS/vm/Set-CmAzIaasUpdateManagement.ps1:33]: revert indentation

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Resources/Project/recoveryvault.yml [hunk 12]
```diff
@@ -1,39 +1,43 @@
...
-    publish:
-      recoveryVault: Iaas-Test-Rec-Vault4
-- name: Iaas-Test-Rec-Vault5
-  backupLevel: Standard
-  location: UK West
-  service:
-    publish:
-      recoveryVault: Iaas-Test-Rec-Vault5
\ No newline at end of file
+  - name: Iaas-Test-Rec-Vault1
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/recoveryvault.yml:14]: done

File: Cloudmarque.Azure/Resources/Project/monitor.yml [hunk 13]
```diff
@@ -29,103 +31,103 @@ alerts:
...
-  functions:
-  itsm:
-  logicApps:
-  notifications:
-  runbooks:
-  sms:
-  voice:
-  webhooks:
-  - serviceUri: https://CoreTestWbSvUri
+  - name: Core-Test-Ag1
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/monitor.yml:34]: done

File: azure-pipelines.yml (hunk 1 above)
Comment (deepend-dev) [reply azure-pipelines.yml:2]: done

File: Tests/New-CmAzDeployment.Tests.ps1 (hunk 2 above)
Comment (deepend-dev) [reply Tests/New-CmAzDeployment.Tests.ps1:8]: resolved

File: Cloudmarque.Azure/Public/Deploy/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1 (hunk 10 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1:44]: no looks good

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 [hunk 14]
```diff
@@ -0,0 +1,144 @@
+function New-CmAzDeployment {
+
+	<#
+		.Synopsis
+		 Auto-deploys Azure resources depending on the type of component defined in settings.
+
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:6]: yes

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        .Example
+		 New-CmAzDeployment -SettingsFile "c:/directory/virtualmachine.yml" -LocalAdminUsername $adminUser -LocalAdminPassword $adminPassword
+	#>
+
+	[CmdletBinding(SupportsShouldProcess, ConfirmImpact = "Medium")]
+	param(
+		[parameter(Mandatory = $true, ParameterSetName = "Settings File")]
+		[String]$SettingsFile,
+		[parameter(Mandatory = $true, ParameterSetName = "Settings Object")]
+        [Object]$SettingsObject,
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:47]: yes

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+		[SecureString]$LocalAdminUsername,
+		[parameter(Mandatory = $false)]
+		[SecureString]$LocalAdminPassword
+	)
+
+	$ErrorActionPreference = "Stop"
+
+	try {
+
+			if ($SettingsFile -and !$SettingsObject) {
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:59]: corrected

File: Tests/New-CmAzDeployment.Tests.ps1 (hunk 3 above)
Comment (deepend-dev) [reply Tests/New-CmAzDeployment.Tests.ps1:10]: done

File: Tests/New-CmAzDeployment.Tests.ps1 (hunk 4 above)
Comment (deepend-dev) [reply Tests/New-CmAzDeployment.Tests.ps1:35]: done

File: Tests/New-CmAzDeployment.Tests.ps1 (hunk 5 above)
Comment (deepend-dev) [reply Tests/New-CmAzDeployment.Tests.ps1:19]: done

File: Tests/New-CmAzDeployment.Tests.ps1 (hunk 6 above)
Comment (deepend-dev) [reply Tests/New-CmAzDeployment.Tests.ps1:23]: for further tests to continue even if the some of the tests fail

File: Tests/New-CmAzDeployment.Tests.ps1 (hunk 7 above)
Comment (deepend-dev) [reply Tests/New-CmAzDeployment.Tests.ps1:27]: done

File: Cloudmarque.Azure/Public/Deploy/Security/security-centre/Set-CmAzSecurityCentre.ps1 (hunk 8 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/Security/security-centre/Set-CmAzSecurityCentre.ps1:45]: yes

File: Cloudmarque.Azure/Public/Deploy/Security/managed-services/New-CmAzSecurityPartner.ps1 (hunk 9 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/Security/managed-services/New-CmAzSecurityPartner.ps1:34]: yes

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/Set-CmAzIaasUpdateManagement.ps1 (hunk 11 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/Set-CmAzIaasUpdateManagement.ps1:33]: there is none

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 (hunk 14 above)
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:6]: Synopsis should be a brief overview, description needs to be in-depth explanation of what the cmdlet does.

[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:6]: Make sure the text under the documentation text is aligned with their headers

File: Cloudmarque.Azure/Public/Deploy/Security/managed-services/New-CmAzSecurityPartner.ps1 [hunk 17]
```diff
@@ -31,14 +31,16 @@
...
-                $SettingsObject = Get-CmAzSettingsFile -Path $SettingsFile
-            }
-            elseif (!$SettingsFile -and !$SettingsObject) {
-                Write-Error "No valid input settings." -Category InvalidArgument -CategoryTargetName "SettingsObject"
-            }
+		Get-InvocationInfo -CommandName $MyInvocation.MyCommand.Name
+
+		if ($PSCmdlet.ShouldProcess((Get-CmAzSubscriptionName), "Deploy managed service security groups.")) {
+
+			if ($SettingsFile -and !$SettingsObject) {
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/Security/managed-services/New-CmAzSecurityPartner.ps1:38]: Looks like this is more indented in relation to its containing if statement

File: Cloudmarque.Azure/Resources/Project/monitor.yml [hunk 18]
```diff
@@ -35,8 +37,8 @@ actionGroups:
     publish:
       actionGroup: Core-Test-Ag1
   armRoles:
-  emails: 
-  - emailAddress: testActionGroup1@Cloudmarque.com
+  emails:
+    - emailAddress: testActionGroup1@Cloudmarque.com
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/monitor.yml:41]: Undo indentation

File: Cloudmarque.Azure/Resources/Project/monitor.yml [hunk 19]
```diff
@@ -52,9 +54,9 @@ actionGroups:
   armRoles:
   emails:
   functions:
-  - functionAppResourceId: Core-Test-Func-Res-Id
-    httpTriggerUrl: https://cloudmarqueTest.com
-    functionName: Core-Test-Func-Name
+    - functionAppResourceId: Core-Test-Func-Res-Id
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/monitor.yml:57]: Undo indentation

File: Cloudmarque.Azure/Resources/Project/monitor.yml [hunk 20]
```diff
@@ -87,8 +89,8 @@ actionGroups:
   functions:
   itsm:
   logicApps:
-  - resourceId: Core-Test-Logic-ResId
-    callbackUrl: https://CoreTestLogicCbUrl
+    - resourceId: Core-Test-Logic-ResId
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/monitor.yml:92]: Undo indentation

File: Cloudmarque.Azure/Resources/Project/monitor.yml [hunk 21]
```diff
@@ -125,7 +127,7 @@ actionGroups:
   sms:
   voice:
   webhooks:
-  - serviceUri: https://CoreTestWbSvUri
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/monitor.yml:128]: Undo indentation

File: Tests/New-CmAzDeployment.Tests.ps1 (hunk 6 above)
Comment (tm-luke) [reply Tests/New-CmAzDeployment.Tests.ps1:24]: Does it still highlight every test that has failed with the continue?

File: Tests/New-CmAzDeployment.Tests.ps1 [hunk 22]
```diff
@@ -0,0 +1,40 @@
+Sync-CloudmarqueAzure
+
+$excludes = @("generators.yml", "naming.yaml")
+$resourceFiles = Get-ChildItem -Path "$PSScriptRoot\..\Cloudmarque.Azure\Resources\Project" -Recurse -Force -Exclude $excludes -File | Select-String "component" -List | Select-Object Filename, Path
+$errorMessage = "Please provide appropriate component value in settings file or object.."
+
+Describe "New-CmAzDeployment Tests" {
+
+    Context "Cmdlet  is successfully identified" {
```
Comment (tm-luke) [inline Tests/New-CmAzDeployment.Tests.ps1:9]: Remove unneeded space.

File: azure-pipelines.yml [hunk 23]
```diff
@@ -26,156 +26,152 @@ trigger:
...
-        workingDirectory: $(Build.SourcesDirectory)
-        pwsh: true
-        failOnStderr: true
-        errorActionPreference: Stop
-        targetType: Inline
-        script: |
-          . .\$(moduleName)\install-dependencies.ps1
-          . .\init.ps1      
-          Sync-CloudmarqueAzure
+    - job:
```
Comment (tm-luke) [inline azure-pipelines.yml:40]: Undo indentation

File: azure-pipelines.yml [hunk 24]
```diff
@@ -26,156 +26,152 @@ trigger:
...
-        script: |
-          . .\init.ps1
-          Test-CloudmarqueAzure
-    - task: PublishTestResults@2
-      displayName: Publish Tests
-      inputs:
-        testResultsFormat: NUnit
-        testResultsFiles: $(moduleName)/publish/tests/*.xml
-        failTaskOnFailedTests: true
+    - job:
```
Comment (tm-luke) [inline azure-pipelines.yml:58]: Undo changes

File: azure-pipelines.yml [hunk 25]
```diff
@@ -26,156 +26,152 @@ trigger:
...
-          . .\init.ps1
-          . .\$(moduleName)\install-dependencies.ps1
-          New-CloudmarqueAzureDocs
-    - task: PublishPipelineArtifact@1
-      displayName: Publish Artifact
-      inputs:
-        targetPath: publish/docs
-        artifactName: docs
-        publishLocation: Pipeline
+    - job:
```
Comment (tm-luke) [inline azure-pipelines.yml:84]: Undo indentation

File: azure-pipelines.yml [hunk 26]
```diff
@@ -26,156 +26,152 @@ trigger:
...
-
-          Write-Host "Build"
-          Build-CloudmarqueAzure -Version $(version) -Suffix $suffix -ReleaseNotes '${{ parameters.ReleaseNotes }}'
-    - task: PublishPipelineArtifact@1
-      displayName: Publish Module Artifact
-      inputs:
-        targetPath: $(moduleName)
-        artifactName: $(moduleName)
-        publishLocation: Pipeline
+    - job:
```
Comment (tm-luke) [inline azure-pipelines.yml:110]: Undo changes

File: azure-pipelines.yml [hunk 27]
```diff
@@ -26,156 +26,152 @@ trigger:
...
-              script: |
-                . .\$(moduleName)\install-dependencies.ps1
-                
-                Publish-Module `
-                  -Path $(moduleName) `
-                  -NuGetApiKey $(NugetApiKey) `
-                  -Repository PSGallery `
-                  -SkipAutomaticTags `
-                  -Force
+    - deployment:
```
Comment (tm-luke) [inline azure-pipelines.yml:143]: Undo changes

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Public/Deploy/Security/managed-services/New-CmAzSecurityPartner.ps1 (hunk 17 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/Security/managed-services/New-CmAzSecurityPartner.ps1:38]: looks fine to me

File: Cloudmarque.Azure/Resources/Project/monitor.yml (hunk 18 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/monitor.yml:41]: done

File: Cloudmarque.Azure/Resources/Project/monitor.yml (hunk 19 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/monitor.yml:57]: done

File: Cloudmarque.Azure/Resources/Project/monitor.yml (hunk 20 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/monitor.yml:92]: done

File: Cloudmarque.Azure/Resources/Project/monitor.yml (hunk 21 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/monitor.yml:128]: done

File: Tests/New-CmAzDeployment.Tests.ps1 (hunk 22 above)
Comment (deepend-dev) [reply Tests/New-CmAzDeployment.Tests.ps1:9]: done

File: azure-pipelines.yml (hunk 23 above)
Comment (deepend-dev) [reply azure-pipelines.yml:40]: done

File: azure-pipelines.yml (hunk 24 above)
Comment (deepend-dev) [reply azure-pipelines.yml:58]: done

File: azure-pipelines.yml (hunk 25 above)
Comment (deepend-dev) [reply azure-pipelines.yml:84]: done

File: azure-pipelines.yml (hunk 27 above)
Comment (deepend-dev) [reply azure-pipelines.yml:143]: done

File: azure-pipelines.yml (hunk 26 above)
Comment (deepend-dev) [reply azure-pipelines.yml:110]: done

File: Cloudmarque.Azure/Public/Deploy/Core/New-CmAzCore.ps1 [hunk 28]
```diff
@@ -45,7 +45,9 @@ function New-CmAzCore {
 
     try {
 
-        if ($PSCmdlet.ShouldProcess((Get-CmAzSubscriptionName), "Deploy Core Monitoring and Logging")) {
+        Get-InvocationInfo -CommandName $MyInvocation.MyCommand.Name
+
+		if ($PSCmdlet.ShouldProcess((Get-CmAzSubscriptionName), "Deploy Core Monitoring and Logging")) {
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/Core/New-CmAzCore.ps1:50]: no

File: Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1 (hunk 14 above)
[the remark this answers is not in the document]
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/New-CmAzDeployment.ps1:6]: Wrapper function to auto-identify cmdlet to based on provided settings and deploy Azure infrastructure.

**Review by tm-luke — APPROVED** [review]

**Pull request merged**